### PR TITLE
Windows: Non config changes to support Gitlab CI

### PR DESCRIPTION
**Pull request by johnwparent**
Adds non-config changes required to run Spack's Gitlab CI workflow with Windows runners. 

Adds generally improved support for escaping Windows escape characters
Adds improved handling of Windows paths and quoting in commands executed by CI
Adds general improvements to the cmd/ci module to support Windows paths and escaping.
Refactors Spack's gpg module to use the filesystem module's `getuid`.
Misc SFN changes not covered by the original SFN PR.

### CODE REVIEW

**Comment by johnwparent** [comment]
CC @kwryankrattiger

File: var/spack/repos/builtin/packages/libxml2/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/libxml2/package.py:249]: (for a future PR) Ideally we could decorate functions like `Stage.source_path` and `Prefix.__str__` to do this `windows_sfn` logic automatically. That can be a follow-up (but IMO a high priority one)

File: lib/spack/spack/ci.py
Comment (scheibelp) [inline lib/spack/spack/ci.py]: (question) Is this quoting important for cases with spaces on Windows, or more generally?

File: lib/spack/llnl/util/filesystem.py
Comment (scheibelp) [inline lib/spack/llnl/util/filesystem.py]: This looks like a bad reference

File: lib/spack/spack/cmd/ci.py
Comment (kwryankrattiger) [inline lib/spack/spack/cmd/ci.py]: Some straggling debug here.

File: lib/spack/spack/environment/environment.py
Comment (kwryankrattiger) [inline lib/spack/spack/environment/environment.py]: I think this is fine for CI, but I think any single letter is allowed, so this should include c-z.

File: lib/spack/llnl/util/filesystem.py
Comment (kwryankrattiger) [inline lib/spack/llnl/util/filesystem.py:201]: Is it okay that this isn't numeric?

**Review by kwryankrattiger — COMMENTED** [review]
Looks pretty good, thanks!

File: lib/spack/spack/environment/environment.py
Comment (scheibelp) [reply lib/spack/spack/environment/environment.py]: I think "c" being interpreted as the scheme is technically a bug, and ideally we'd wrap urlparse with something that can understand this on Windows.

I think a TODO would be helpful, if we don't change that here.

File: lib/spack/llnl/util/filesystem.py
Comment (scheibelp) [reply lib/spack/llnl/util/filesystem.py:201]: I don't think we do anything with it that requires it to be a number, although there might be confusion if there is the possibility of changing between `"1"` and `1`, for example.

**Review by scheibelp — CHANGES_REQUESTED** [review]
Some preliminary comments. I can look at this more thoroughly tomorrow.

File: lib/spack/llnl/util/filesystem.py
Comment (johnwparent) [reply lib/spack/llnl/util/filesystem.py]: I don't know how many times I can fix something in a rebase conflict and still have it propagate to the final changes.... 😬 :💢 
Nice catch

File: var/spack/repos/builtin/packages/libxml2/package.py
Comment (johnwparent) [reply var/spack/repos/builtin/packages/libxml2/package.py:249]: SFN handling has enough weird corner cases/behavior (like the filesystem explicitly needing to enable it, etc) that IMO, we should limit its application to only the programs that don't have any better handling for long paths or paths with spaces. To the best of my knowledge this is only NMake and MSBuild (CMake, scons, meson, setup-tools, etc all handle spaces + long paths perfectly fine). That said maybe this is something we want to add to handle cases where users can't enable long path support in general. I think this warrants a larger conversation and should probably be added to my PR where I add the system checks for long paths, symlinks, etc to the CLI.

File: lib/spack/llnl/util/filesystem.py
Comment (johnwparent) [reply lib/spack/llnl/util/filesystem.py:201]: It's fine this isn't numeric, that said Peter makes a good point...
We need to revisit this and similar concepts in general to use Windows ACLS and UIDs so if this isn't causing errors I'd like to punt to that effort. That said @scheibelp I'll defer to your judgement here.

File: lib/spack/spack/environment/environment.py
Comment (johnwparent) [reply lib/spack/spack/environment/environment.py]: IIRC we tried to fix this last summer and ended up uncovering a bit of a hornet's nest w.r.t. url parsing on Windows. I'll check into Markus' old PRs and see if I can pull some of that effort in here, otherwise will genericize using our drive letter regex.

File: lib/spack/spack/cmd/ci.py
Comment (johnwparent) [reply lib/spack/spack/cmd/ci.py]: Nice catch

File: lib/spack/spack/ci.py
Comment (johnwparent) [reply lib/spack/spack/ci.py]: It's a combination of spaces and other characters that would otherwise cause the argument parsing to decompose the url into useless gibberish. I can't recall the exact issue off the top of my head but there was more to it than spaces. 
That said, all issues were due to Windows caveats.

File: lib/spack/llnl/util/filesystem.py
Comment (kwryankrattiger) [reply lib/spack/llnl/util/filesystem.py:201]: I thought I saw a some places where we check `if fs.getuid() == 0` which I don't know if that is generally allowed comparing string vs int.

I have python 3.8 locally and it seems fine, even `"0" == 0` yields `False` which I think is correct. I don't see anywhere where we try to do `int(fs.getuid())` which I think is the more problematic case as you pointed out.

Comment (scheibelp) [reply lib/spack/llnl/util/filesystem.py:201]: This function still returns `0` (not `"0"`) in all cases where the user is an admin. I did `grep -i "getuid" -r lib/spack/spack`:

1. I mostly see this in tests testsing for `== 0` (so those are fine)
2. I see a check in `Stage` comparing this to `get_owner_uid` (so possibly this is an improvement?)
3. One use in `util.gpg` that calls `str()`  as part of making a storage path (so that is fine)

Overall except for [2] these are definitely fine (documentation on how we might return different types could be useful (or we could encode that as a mypy return type). We might also want to add documentation discussing whether/how the behavior of `getuid` and `get_owner_uid` match up.

File: lib/spack/spack/cmd/ci.py
Comment (scheibelp) [inline lib/spack/spack/cmd/ci.py]: A few things:

* I recommend calling this `unicode_escape` vs. `encode` (what it's returning is the result of calling `decode`, so calling it encode is distracting IMO, though probably is not strictly incorrect depending on what you mean by `encode`)
* Explicitly call `.decode('utf-8')` (that's the default, but I think it's a useful thing to emphasize in this context)
* Add a docstring like "transform string, replacing each unicode character with a sequence of characters representing the unicode code point"

File: lib/spack/spack/environment/environment.py
Comment (scheibelp) [reply lib/spack/spack/environment/environment.py]: I don't want to overextend this, so I'll say you can take a partial approach if you like (or add a TODO).

**Review by scheibelp — DISMISSED** [review]
Additional request for helper function. Overall changes look good.

File: lib/spack/llnl/util/filesystem.py
Comment (kwryankrattiger) [reply lib/spack/llnl/util/filesystem.py:201]: Nice, I am fine then.

File: lib/spack/spack/environment/environment.py
Comment (johnwparent) [reply lib/spack/spack/environment/environment.py]: I was thinking of Harmen's PR https://github.com/spack/spack/pull/34452
The helper methods in here should fix this

Comment (johnwparent) [reply lib/spack/spack/environment/environment.py]: ~~On second thought, I think this indicates a more fundamental problem. Why are we passing valid file urls in one context and raw Windows paths in another? The Windows paths should also be formatted as file urls so they can be picked up properly by url parse. If we were just passing raw paths on *nix platforms, we'd get `''` which would also fail here.~~
~~Going to look into where this is coming from and fix it there.~~
Misread some code, back to handling this here.

File: lib/spack/spack/util/url.py
Comment (kwryankrattiger) [inline lib/spack/spack/util/url.py]: Are the drive names `A` and `B` allowed? I thought they started at `C` for some historical reasons.

Comment (johnwparent) [reply lib/spack/spack/util/url.py]: They're legal drive names, historically they were reserved for floppy drives but I don't think that's technically enforced anymore.

Comment (kwryankrattiger) [reply lib/spack/spack/util/url.py]: Okay, even if it is enforced, I don't think there is anything saying we can't install spack packages to a floppy.

File: lib/spack/spack/cmd/ci.py
Comment (johnwparent) [reply lib/spack/spack/cmd/ci.py]: Makes a lot of sense, amended.

File: lib/spack/spack/environment/environment.py
Comment (johnwparent) [reply lib/spack/spack/environment/environment.py]: This has been handled by relying on scheme validation rather than a hard check for Windows drive labels.

File: lib/spack/spack/ci.py
Comment (kwryankrattiger) [inline lib/spack/spack/ci.py]: This extra double quoting seems to be broken on windows 😞 What ends up in the generated `install.sh` is a bunch of commands with `""asdfasdfasdf ()""` which is incorrect.

I know we did this for windows do handle things not being quoted by default. We may just need to have a `quote_on_windows` function here.

**Comment by kwryankrattiger** [comment]
+2 from me. It looks like libxml2 built and we testee concretize, environments, install touched by this change. Unit tests look fine to me, the real test of this will be once we get Windows CI merged (#43967).

@scheibelp I will plan on merging this tomorrow morning unless you have anything else that you think is blocking this.

**Review by kwryankrattiger — APPROVED** [review]

**Pull request merged**
